### PR TITLE
expression, util: consider unsigned when cast

### PR DESCRIPTION
**Pull request by XuHuaiyu**
This commit improves the `castIntAsXXSig.evalXX()` and `castXXAsIntSig.evalInt()`,
we should consider unsigned int and signed int individually when casting,
and we should round but not truncate when converting a float to int.

PTAL @hanfei1991 @shenli

### CODE REVIEW

**Comment by shenli** [comment]
Any test case?

**Comment by hanfei1991** [comment]
If a column is evaluated as int, should we consider it's unsinged flag?

**Comment by XuHuaiyu** [comment]
@hanfei1991 
You mean that a column is not int, but we evaluate it as int?
In this case, there will be no unsigned flag problem.

**Comment by XuHuaiyu** [comment]
PTAL @shenli

File: expression/builtin_cast_test.go [hunk 1]
```diff
@@ -931,6 +937,40 @@ func (s *testEvaluatorSuite) TestWrapWithCastAsTypesClasses(c *C) {
...
+
+	unsignedIntExpr := &Column{RetType: &types.FieldType{Tp: mysql.TypeLonglong, Flag: mysql.UnsignedFlag}, Index: 0}
+
+	// test cast unsigned int as string.
+	strExpr, err := WrapWithCastAsString(unsignedIntExpr, ctx)
+	c.Assert(err, IsNil)
+	c.Assert(strExpr.GetTypeClass(), Equals, types.ClassString)
+	strRes, isNull, err := strExpr.EvalString([]types.Datum{types.NewUintDatum(math.MaxUint64)}, sc)
+	// overflow
+	c.Assert(err, NotNil)
```
Comment (shenli) [inline expression/builtin_cast_test.go]: Update this after rebase master.

**Comment by XuHuaiyu** [comment]
@shenli  PTAL

File: expression/builtin_cast_test.go [hunk 2]
```diff
@@ -23,6 +23,7 @@ import (
 	"github.com/pingcap/tidb/util/charset"
 	"github.com/pingcap/tidb/util/testleak"
 	"github.com/pingcap/tidb/util/types"
+	"math"
```
Comment (shenli) [inline expression/builtin_cast_test.go]: Move this to the above position.

File: expression/builtin_cast.go [hunk 3]
```diff
@@ -756,7 +822,9 @@ func (b *builtinCastDurationAsIntSig) evalInt(row []types.Datum) (res int64, isN
 	if isNull || err != nil {
 		return res, isNull, errors.Trace(err)
 	}
-	res, err = val.ToNumber().ToInt()
+	dec := val.ToNumber()
```
Comment (shenli) [inline expression/builtin_cast.go]: Add test case for this.

**Comment by XuHuaiyu** [comment]
1. add some test cases for converting REAL/ DECIMAL/ TIME/ DURATION to INT where existing ROUND
2. fix #3470

**Comment by XuHuaiyu** [comment]
PTAL @shenli @hanfei1991

**Comment by XuHuaiyu** [comment]
PTAL @shenli @hanfei1991

**Comment by hanfei1991** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**